# Hola &#x1F600;,

Soy **Hesus Garcia** – **"Soy el único Hesus que conoces (y probablemente conocerás) 🌟"** – Sí, como "Jesús", pero con una H que me hace único. Puede sonar raro, pero créeme, ¡no lo olvidarás! Como tu revisor en Triple-Ten, estoy aquí para guiarte y ayudarte a mejorar tu código. Si algo necesita un ajuste, no hay de qué preocuparse; ¡aquí estoy para hacer que tu trabajo brille con todo su potencial! ✨

Cada vez que encuentre un detalle importante en tu código, te lo señalaré para que puedas corregirlo y así te prepares para un ambiente de trabajo real, donde el líder de tu equipo actuaría de manera similar. Si en algún momento no logras solucionar el problema, te daré más detalles para ayudarte en nuestra próxima oportunidad de revisión.

Es importante que cuando encuentres un comentario, **no los muevas, no los modifiques, ni los borres**.

---

### Formato de Comentarios

Revisaré cuidadosamente cada implementación en tu notebook para asegurar que cumpla con los requisitos y te daré comentarios de acuerdo al siguiente formato:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> - ¡Excelente trabajo! Esta parte está bien implementada y contribuye significativamente al análisis de datos o al proyecto. Continúa aplicando estas buenas prácticas en futuras secciones.
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Este código está correcto, pero se puede optimizar. Considera implementar mejoras para que sea más eficiente y fácil de leer. Esto fortalecerá la calidad de tu proyecto.
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error en el código que es necesario corregir para aprobar esta sección. Por favor, revisa y corrige este punto, ya que es fundamental para la validez del análisis y la precisión de los resultados.
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor** que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto.
- **Áreas de mejora:** Sugerencias sobre aspectos donde puedes mejorar.
- **Temas adicionales para investigar:** Ideas de temas opcionales que puedes explorar por tu cuenta para desarrollar aún más tus habilidades.

Estos temas adicionales no son obligatorios en esta etapa, pero pueden serte útiles para profundizar en el futuro.

---


Esta estructura en viñetas facilita la lectura y comprensión de cada parte del comentario final.

También puedes responderme de la siguiente manera si tienes alguna duda o quieres aclarar algo específico:


<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
    
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
    
</div>


**¡Empecemos!** &#x1F680;


# Proyecto RappiPlus: de datos a decisiones de negocio

**Introducción**


El objetivo de este proyecto es evaluar el desempeño del servicio **RappiPlus** para apoyar **decisiones de negocio basadas en datos**.

Se trabajan con múltiples datasets del negocio:

- **rappiplus_orders_raw.csv** → información de pedidos, precios, descuentos y revenue  
- **rappiplus_catalog.csv** → costos de productos, categorías y proveedores  
- **rappiplus_marketing_spend.csv** → inversión en marketing por canal y país  
- **events / users / user_activity (SQL)** → comportamiento del usuario dentro de la plataforma  
- **experiment_checkout_ui.csv** → resultados de un experimento A/B en el checkout  

El análisis sigue una lógica clara y progresiva:

1. 🔍 Evaluar si podemos confiar en los datos (calidad de datos en Python) 

2. 💰 Analizar si el negocio es rentable (revenue, costos y profit)  

3. 🛒 Entender dónde se pierden los usuarios (funnel de conversión)  

4. 🔁 Evaluar si los usuarios regresan (retención por cohortes)  

5. 🧪 Validar si los cambios generan impacto (test estadístico)  

6. 📊 Comunicar los resultados (dashboard en BI)  

A lo largo del proyecto, se transforman datos en insights para responder preguntas clave del negocio y proponer **recomendaciones accionables**.

---

## 🔹 Paso 1: Cargar y validar la calidad de los datos

---

### 1.1 Carga de datos y vista rápida

**🎯 Objetivo:** Familiarizarte con la estructura de los datasets del negocio antes de analizarlos.

**Instrucciones:**

- Importa las librerías necesarias
- Carga los archivos:
  - `rappiplus_orders_raw.csv`
  - `rappiplus_catalog.csv`
  - `rappiplus_marketing_spend.csv`
- Guarda los DataFrames en:
  - `orders`, `catalog`, `marketing`
- Explora cada dataset.
---

In [1]:
# importar librerías
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt 
from sqlalchemy import create_engine  # Para conexión SQL
from scipy import stats  # Para test estadístico

In [2]:
# cargar archivos
orders =  pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_orders_raw.csv') # tu código aquí
catalog =  pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_catalog.csv')# tu código aquí
marketing =  pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_marketing_spend.csv') # tu código aquí

In [3]:
# Creamos un punto de restauración
work_orders = orders.copy()
work_catalog = catalog.copy()
work_marketing = marketing.copy()

In [4]:
# explorar datasets: Orders
# tu código aquí
work_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25100 entries, 0 to 25099
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id_pedido           25100 non-null  object 
 1   id_usuario          25100 non-null  object 
 2   fecha_hora_pedido   25100 non-null  object 
 3   pais                24800 non-null  object 
 4   dispositivo         25080 non-null  object 
 5   fuente_referencia   25070 non-null  object 
 6   nombre_producto     25070 non-null  object 
 7   categoria_producto  25020 non-null  object 
 8   cantidad            25050 non-null  float64
 9   precio_unitario     25050 non-null  float64
 10  monto_descuento     25050 non-null  float64
 11  monto_total         25100 non-null  float64
dtypes: float64(4), object(8)
memory usage: 2.3+ MB


In [5]:
# explorar datasets: Orders
# tu código aquí
work_orders.describe()

,cantidad,precio_unitario,monto_descuento,monto_total
count,25050.000000,25050.000000,25050.000000,2.510000e+04
mean,7.092735,259.305549,4.500798,2.072680e+03
std,296.277003,138.726461,5.223010,9.894995e+04
min,-2.000000,20.030000,0.000000,-4.926500e+02
25%,1.000000,138.377500,0.000000,1.805075e+02
50%,2.000000,258.715000,0.000000,3.417500e+02
75%,2.000000,380.332500,10.000000,5.185800e+02
max,20000.000000,499.960000,15.000000,8.840200e+06


In [6]:
# explorar datasets: Orders
# tu código aquí
work_orders.head()

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
0,order_0,user_6993,2025-05-22,Argentina,desktop,organic,Jacket-Winter-M,Moda,2.0,332.69,0.0,665.37
1,order_1,user_1329,2025-06-15,Mexico,desktop,paid_search,Tablet-Standard-64GB,Electronica,1.0,176.86,5.0,171.86
2,order_2,user_3194,2025-05-02,Argentina,desktop,social,Blender-XL-Red,Hogar,2.0,102.99,10.0,195.99
3,order_3,user_4510,2025-06-09,Colombia,mobile,social,Tablet-Standard-64GB,Electronica,1.0,257.87,15.0,242.87
4,order_4,user_5044,2025-03-30,Argentina,desktop,paid_search,Blender-XL-Red,Hogar,1.0,336.28,0.0,336.28


Información general del dataset: **ORDERS**
25,100 pedidios registrados. 12 Columnas
Problemas de calidad detectados: 
🔴 Problemas críticos: País con 300 valores faltanes (25,100-24800), categoría_producto: 80 valores faltantes (25,100- 25,020)
🟡 Problemas moderados: dispositivo con 20 valores faltantes, nombre_producto con 30 valores faltantes, cantidad con 50 valores faltantes, precio_unitario con 50 valores faltantes, monto_descuento con 50 valores faltantes. 
✅ Columnas sin valores faltantes: id_pedido, id_usuario, y monto_total.

Tipos de datos: 
object: Texto (fechas, nombres, categorías) Fecha como texto, tipo incorrecto. Cambiar a DATE.
float64: Números decimales (precios, cantidades, montos).
Estadísticas descriptivas:
Min: -2.00 ¿comprar -2 productos? Se verifica y hay 4 ordenes consecutivas order_266 a la order_269 con números negativos en cantidad y en monto, puede ser que hayan sido devoluciones, o equivocaciones en la marcación o error en el sistema, los cuatro pedidos son de fechas distintas, con descuento diferente o sin, todos están sobre el mismo producto Phone-Pro-128GB 
Max: 20,000 ¿compra de 20,000 en una orden?
Monto total negativo: min:-$492.65
Valores extremos sospechosos: Max monto_total:$8,840,200 (¿orden de casi 9 millones?)
Desviación estándar muy alta en cantidad (296) y monto_total(98,949) 
Indagar si son errores del sistema o casos especiales válidos.

In [7]:
# explorar datasets: Catalog
# tu código aquí
work_catalog.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   nombre_producto     7 non-null      object 
 1   categoria_producto  7 non-null      object 
 2   costo_unitario      7 non-null      float64
 3   proveedor           7 non-null      object 
dtypes: float64(1), object(3)
memory usage: 352.0+ bytes


In [8]:
# explorar datasets: Catalog
# tu código aquí
work_catalog.describe()

,costo_unitario
count,7.000000
mean,102.252857
std,111.011563
min,10.120000
25%,16.905000
50%,25.210000
75%,182.975000
max,280.680000


In [9]:
# explorar datasets: Catalog
# tu código aquí
work_catalog.head()

,nombre_producto,categoria_producto,costo_unitario,proveedor
0,Laptop-Gaming-16GB,Electrónica,280.68,"Fuller, Pena and Myers"
1,Phone-Pro-128GB,Electrónica,10.12,King Ltd
2,Tablet-Standard-64GB,Electrónica,25.21,Bowers LLC
3,Blender-XL-Red,Hogar,176.64,Long-Reid
4,Vacuum-Pro-Black,Hogar,16.60,"Rivera, Carr and Finley"


Información general del dataset: **CATALOGO**
Catalog: 7 filas de datos. Los índices van de 0 a 6. 4 Columnas en total. 

📋 Información de cada columna:
nombre_producto: 7 valores no nulos, tipo object (texto)
categoria_producto: 7 valores no nulos, tipo object (texto)
costo_unitario: 7 valores no nulos, tipo float64 (números decimales)
proveedor: 7 valores no nulos, tipo object (texto)

📊 Análisis de las estadísticas del costo_unitario
count = 7.000000
Tiene exactamente 7 productos en el catálogo, todos con información de costo unitario completa.
mean = 102.252857
El costo promedio de tus productos es de $102.25. Este es el punto de referencia central para entender si un producto es "caro" o "barato" de producir.
std = 111.011563
La desviación estándar es alta ($111.01), lo que indica que hay mucha variabilidad en los costos. Algunos productos son muy baratos de producir y otros muy costosos.
min = 10.120000 y max = 280.680000

Hay una gran diferencia entre el producto más barato ($10.12) y el más caro ($280.68) de producir. ¡Una diferencia de casi $270!
Percentiles (25%, 50%, 75%)
25% (Q1) = $16.91: El 25% de tus productos cuesta menos de $16.91 producir
50% (Q2/Mediana) = $25.21: La mitad de tus productos cuesta menos de $25.21 producir
75% (Q3) = $182.98: El 75% de tus productos cuesta menos de $182.98 producir
Parece que se tienen dos grupos de productos:
Productos de bajo costo (electrónicos como Phone y Tablet): ~$10-25
Productos de alto costo (Laptop y algunos productos de hogar): ~$177-281

In [10]:
# explorar datasets: Marketing
# tu código aquí
work_marketing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1620 entries, 0 to 1619
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   fecha       1620 non-null   object 
 1   pais        1620 non-null   object 
 2   id_campaña  1620 non-null   object 
 3   canal       1519 non-null   object 
 4   gasto       1620 non-null   float64
dtypes: float64(1), object(4)
memory usage: 63.4+ KB


In [11]:
# explorar datasets: Marketing
# tu código aquí
work_marketing.describe()

,gasto
count,1620.00000
mean,1772.74292
std,734.43294
min,501.11000
25%,1128.03000
50%,1782.42500
75%,2420.68500
max,2999.36000


In [12]:
# explorar datasets: Marketing
# tu código aquí
work_marketing.head()

,fecha,pais,id_campaña,canal,gasto
0,2025-01-01,Mexico,organic_Mexico,organic,2446.25
1,2025-01-01,Mexico,paid_search_Mexico,paid_search,2704.34
2,2025-01-01,Mexico,social_Mexico,social,2045.01
3,2025-01-01,Colombia,organic_Colombia,organic,2597.21
4,2025-01-01,Colombia,paid_search_Colombia,paid_search,1771.40


Información general del dataset **MARKETING**:
RangeIndex: 1620 entries, 0 to 1619
El dataset tiene 1,620 filas (registros de inversión en marketing)
Las filas están numeradas del 0 al 1619
Data columns (total 5 columns):

📋 Análisis por columna:
fecha: 1620 valores (sin datos faltantes) - tipo object, necesita estar en DATE.
pais: 1620 valores (sin datos faltantes) - tipo object
id_campaña: 1620 valores (sin datos faltantes) - tipo object
canal: 1519 valores (101 datos faltantes) - tipo object
gasto: 1620 valores (sin datos faltantes) - tipo float64
⚠️ Problema detectado:
La columna canal tiene 101 valores faltantes (1620 - 1519 = 101). Esto significa que hay campañas sin información del canal de marketing.

📊 Desglose de las estadísticas:
count: 1620.00000
Tienes 1,620 registros de inversión en marketing (sin datos faltantes)
mean: $1772.74292. El gasto promedio por campaña es $1,772.74
std: $734.43294. La desviación estándar es $734.43, indicando variabilidad moderada en los gastos
min: $501.11000. La inversión mínima registrada fue $501.11
25%: $1128.03000. El 25% de las campañas gastó $1,128.03 o menos
50%: $1782.42500. La mediana es $1,782.43 (50% gasta menos, 50% gasta más)
75%: $2420.68500. El 75% de las campañas gastó $2,420.69 o menos
max: $2999.36000. La inversión máxima fue $2,999.36

🔍 Interpretación clave:
La media ($1,772.74) y la mediana ($1,782.43) están muy cercanas, lo que indica una distribución bastante equilibrada sin valores extremos que distorsionen el promedio.
El rango va de $501 a $2,999, mostrando una estrategia de inversión diversificada con campañas de diferentes escalas.

📊 Estructura del dataset de marketing:
Esta tabla registra la inversión diaria por campaña en diferentes países y canales. Cada fila representa el gasto de una campaña específica en una fecha determinada.
📅 fecha: La fecha de la inversión. Todas estas campañas empezaron el 1 de enero de 2025
🌎 pais: El país donde se ejecutó la campaña. (mercados principales de RappiPlus)
🏷️ id_campaña: Identificador único de cada campaña. Sigue el patrón: [tipo_canal]_[país]
Ejemplos: "organic_Mexico", "paid_search_Colombia"
📢 canal: El tipo de canal de marketing utilizado: organic: Marketing orgánico (SEO, contenido)
paid_search: Búsqueda pagada (Google Ads), social: Redes sociales (Facebook, Instagram)
💰 gasto: La inversión para esa campaña, Rango de $1,771 a $2,704 en estos ejemplos


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - La exploración inicial de los datasets está desarrollada de manera ordenada y facilita comprender rápidamente la estructura de la información disponible. Se identifican correctamente valores faltantes, tipos de datos inconsistentes y registros atípicos que podrían impactar el análisis posterior. Además, la interpretación de los resultados muestra una buena capacidad para relacionar los hallazgos técnicos con posibles situaciones reales del negocio, especialmente en la detección de órdenes sospechosas y variaciones en costos y gastos de marketing.
</div>


---

### Revisión y calidad de datos

**🎯 Objetivo:** Detectar y corregir problemas en los datos que puedan afectar el análisis de revenue, costos y rentabilidad.
Se revisan los 3 datasets
- Validar y convertir fechas al formato correcto  
- Revisar variables numéricas (sin negativos o ceros inválidos)  
- Verificar consistencia de montos  
- Eliminar duplicados  
- Revisar variables categóricas 
---

In [13]:
# tu código aquí
# 🚀 EJECUTAR PIPELINE COMPLETO PARA CADA DATASETS

# Ejecutar pipeline completo
print("\n" + "="*60)
print("🔄 EJECUTANDO PIPELINE COMPLETO ORDERS")
print("="*60)

# 📊 PIPELINE: ANÁLISIS EXPLORATORIO DE DATOS

def pipeline_analisis_exploratorio(dataframe, nombre_dataset):
    """
    Pipeline completo para análisis exploratorio de datos
    """    
    print("📋 1. INFORMACIÓN GENERAL:")
    print(f"   • Filas: {len(dataframe):,}")
    print(f"   • Columnas: {len(dataframe.columns)}")
    
    return {
        'filas': len(dataframe),
        'columnas': len(dataframe.columns),
        'valores_faltantes': dataframe.isnull().sum().sum(),
        'duplicados': work_orders.duplicated().sum()
    }
# Pipeline para Orders
resumen_orders = pipeline_analisis_exploratorio(work_orders, "Orders")

# 🚀 EJECUTAR ANÁLISIS PARA CADA DATASET

# Para work_orders
print("🔍 ANÁLISIS INICIAL:")

print(f"Total datos faltantes:") 
print(f"{work_orders.isna().sum()}")

print(f"Total datos únicos:")
print(f"Países únicos: {work_orders['pais'].nunique()}")
print(f"Productos únicos: {work_orders['nombre_producto'].nunique()}")
print(f"Usuarios únicos: {work_orders['id_usuario'].nunique()}")


print(f"Total datos duplicados:") 
print(f"{work_orders.duplicated().value_counts()}")

# Ver distribuciones
print("\n📊 DISTRIBUCIONES:")
print(work_orders['pais'].value_counts())
print(work_orders['dispositivo'].value_counts())



🔄 EJECUTANDO PIPELINE COMPLETO ORDERS
📋 1. INFORMACIÓN GENERAL:
   • Filas: 25,100
   • Columnas: 12
🔍 ANÁLISIS INICIAL:
Total datos faltantes:
id_pedido               0
id_usuario              0
fecha_hora_pedido       0
pais                  300
dispositivo            20
fuente_referencia      30
nombre_producto        30
categoria_producto     80
cantidad               50
precio_unitario        50
monto_descuento        50
monto_total             0
dtype: int64
Total datos únicos:
Países únicos: 6
Productos únicos: 7
Usuarios únicos: 7642
Total datos duplicados:
False    25000
True       100
dtype: int64

📊 DISTRIBUCIONES:
Colombia     7520
Mexico       7502
Argentina    7291
mexico        865
colombia      823
argentina     799
Name: pais, dtype: int64
desktop    12759
mobile     12321
Name: dispositivo, dtype: int64


**Se procede con la limpieza en work_orders:**
- Cambio de dtype: fecha_hora_pedido a DATE, se renombra a "fecha_pedido", (la hora no es un dato presente en el dataset).
- Convierte valores negativos a positivos en columnas numéricas, principal en lo detectado: (en cantidad -2, -1, -1,-1) y monto total (-192.62, -38.5, -493.65, -423.53.): Son 4 casos en todo el dataset, son distintos id_usuarios, fechas. (O fue error del sistema al registrarlo en negativo al ser consecutivo el error, o pudieron ser devoluciones y se registraron en conjunto; al ser pocos eventos registrados, se procede a mantenerlos en positivo infiriendo error del sistema/captura).
- Estandarizar el nombre de países únicos del dataset son 3 (Mexico, Colombia y Argentina) no 6, se identifica registro incorrecto en el uso de mayúsculas/minúsculas, se integran los nombres registrados en mínuscula.
- Para los duplicados, se eliminan100 DUPLICADOS EXACTOS con: (mismo usuario, producto, fecha, monto, descuento, cantidad, etc). (Se infiere error del sistema en la duplicación en el momento de la compra/ó doble clic por el cliente).
- Imputar valores faltantes en las siguientes categorías:
- país 300, se procede a imputar entre los 3 paises respetando proproción previa existente, distribución aleatoria con probabilidad calculada y manteniendo la consistencia siguiento el mismo %patrón de datos reales.
- Dispositivo 20, fuente_referencia 30, nombre_producto 30, categoria_producto 80, se imputan/identifican como dato desconocido "unknown".
- Para cantidad, se imputa en los 50 faltantes con = 1 cantidad.
- Imputación para el precio_unitario con datos del mismo dataset, precio_unitario = (monto_total + monto_descuento) / cantidad
- Validar consistencia de montos: Resultado= 1,142 inconsistencias en montos detectadas. ¡Son diferencias de centavos! Esto es típico de redondeos automáticos del sistema. El impacto financiero total de todas las inconsistencias es de: $0.48 centavos. Impacto mínimo: $0.48 sobre miles de pedidos es insignificante. Calidad alta: Las inconsistencias son por redondeos, no errores graves. Usar monto_total: Es el valor real que pagó el cliente.
- Validar rangos de precios, detectar outliers con el Método IQR (eliminar precios extremos si se detectan)
   • Q1 (25%): $138.47
   • Q3 (75%): $380.38
   • IQR: $241.91
   • Límite inferior: $-224.39
   • Límite superior: $743.24
   • Outliers detectados: 0
 

In [14]:
#Converir columna de work_order (fecha_hora_pedido) a tipo fecha con pd.to_datetime ()
work_orders['fecha_hora_pedido'] = pd.to_datetime(work_orders["fecha_hora_pedido"], errors= "coerce") 

#Cambiar nombre de columna usando rename ()
work_orders = work_orders.rename(columns={'fecha_hora_pedido': 'fecha_pedido'})

#Verificar el cambio 
work_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25100 entries, 0 to 25099
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   id_pedido           25100 non-null  object        
 1   id_usuario          25100 non-null  object        
 2   fecha_pedido        25100 non-null  datetime64[ns]
 3   pais                24800 non-null  object        
 4   dispositivo         25080 non-null  object        
 5   fuente_referencia   25070 non-null  object        
 6   nombre_producto     25070 non-null  object        
 7   categoria_producto  25020 non-null  object        
 8   cantidad            25050 non-null  float64       
 9   precio_unitario     25050 non-null  float64       
 10  monto_descuento     25050 non-null  float64       
 11  monto_total         25100 non-null  float64       
dtypes: datetime64[ns](1), float64(4), object(7)
memory usage: 2.3+ MB


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - La etapa de limpieza y validación de datos muestra un proceso analítico cuidadoso y bien justificado. Se identifican correctamente inconsistencias, duplicados y problemas de formato, acompañando cada decisión con una interpretación lógica del contexto de negocio. Destaca especialmente la validación de montos y el análisis de outliers, ya que permiten fortalecer la confiabilidad del dataset antes de avanzar hacia métricas de rentabilidad y comportamiento de usuarios.
</div>


In [15]:
# Convierte valores negativos a positivos en columnas numéricas
columnas_numericas = {'cantidad', 'precio_unitario', 'monto_descuento', 'monto_total'}

for col in columnas_numericas:
    if col in work_orders.columns:
        negativos_antes = (work_orders[col] < 0).sum()
        work_orders[col] = work_orders[col].abs()  # Valor absoluto
        print(f"✅ {col}: {negativos_antes} valores negativos corregidos")

# Verificar que los cambios se aplicaron
print("\n🔍 VERIFICACIÓN:")
for col in columnas_numericas:
    if col in work_orders.columns:
        negativos_restantes = (work_orders[col] < 0).sum()
        print(f"   {col}: {negativos_restantes} valores negativos restantes")

# Guardar el dataset limpio
work_orders.to_csv('work_orders_clean.csv', index=False)
print("\n💾 cambio guardado en 'work_orders_clean.csv'")

✅ monto_total: 4 valores negativos corregidos
✅ precio_unitario: 0 valores negativos corregidos
✅ cantidad: 4 valores negativos corregidos
✅ monto_descuento: 0 valores negativos corregidos

🔍 VERIFICACIÓN:
   monto_total: 0 valores negativos restantes
   precio_unitario: 0 valores negativos restantes
   cantidad: 0 valores negativos restantes
   monto_descuento: 0 valores negativos restantes

💾 cambio guardado en 'work_orders_clean.csv'


In [16]:
# Estandarizar nombres de países en work_orders
work_orders['pais'] = work_orders['pais'].str.lower().str.title()
# Verificar el resultado
print("Países únicos después de la limpieza:")
print(work_orders['pais'].value_counts())

# Guardar cambio en el dataset limpio
work_orders.to_csv('work_orders_clean.csv', index=False)
print("\n💾 cambio guardado en 'work_orders_clean.csv'")


Países únicos después de la limpieza:
Mexico       8367
Colombia     8343
Argentina    8090
Name: pais, dtype: int64

💾 cambio guardado en 'work_orders_clean.csv'


In [17]:
def eliminar_duplicados_exactos(df):
    """
    Elimina solo duplicados EXACTOS manteniendo la primera ocurrencia
    """
    print("🔍 Analizando duplicados exactos...")
    # Contar filas antes
    filas_antes = len(df)
    
    # Identificar duplicados exactos en todas las columnas
    duplicados_exactos = df.duplicated(keep='first')
    
    # Mostrar información antes de eliminar
    print(f"📊 Filas antes: {filas_antes}")
    print(f"🔴 Duplicados exactos encontrados: {duplicados_exactos.sum()}")
    
    # Eliminar duplicados exactos (mantener solo la primera ocurrencia)
    df_limpio = df.drop_duplicates(keep='first')
    
    # Contar filas después
    filas_despues = len(df_limpio)
    eliminadas = filas_antes - filas_despues
    
    print(f"✅ Filas después: {filas_despues}")
    print(f"🗑️ Filas eliminadas: {eliminadas}")
    
    return df_limpio

# Aplicar la función al dataset work_orders
work_orders_sin_duplicados = eliminar_duplicados_exactos(work_orders)

# Actualizar la variable work_orders con los datos sin duplicados
work_orders = work_orders_sin_duplicados

# Guardar cambio en el dataset limpio
work_orders.to_csv('work_orders_clean.csv', index=False)
print("\n💾 cambio guardado en 'work_orders_clean.csv'")

🔍 Analizando duplicados exactos...
📊 Filas antes: 25100
🔴 Duplicados exactos encontrados: 100
✅ Filas después: 25000
🗑️ Filas eliminadas: 100

💾 cambio guardado en 'work_orders_clean.csv'


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - El proceso de normalización y limpieza de datos está bien ejecutado y mantiene una secuencia lógica clara. La corrección de valores negativos, la estandarización de categorías y la eliminación de duplicados exactos fortalecen notablemente la calidad del dataset. Además, la verificación posterior a cada transformación aporta mayor confiabilidad y control sobre los cambios realizados en la información.
</div>


In [18]:
# Imputar país faltante usando la distribución proporcional existente
paises_distribucion = work_orders['pais'].value_counts(normalize=True)
print("Distribución actual de países:")
print(paises_distribucion)

# Imputar valores faltantes manteniendo la proporción
np.random.seed(42) 

# Obtener índices de valores faltantes
indices_faltantes = work_orders['pais'].isnull()
n_faltantes = indices_faltantes.sum()

# Generar países aleatorios según la distribución existente
paises_imputados = np.random.choice(
    paises_distribucion.index,
    size=n_faltantes,
    p=paises_distribucion.values
)

# Aplicar la imputación 
work_orders.loc[indices_faltantes, 'pais'] = paises_imputados
print(f"✅ {n_faltantes} valores de país imputados")

# Guardar cambio en el dataset limpio
work_orders.to_csv('work_orders_clean.csv', index=False)
print("\n💾 cambio guardado en 'work_orders_clean.csv'")

Distribución actual de países:
Mexico       0.337692
Colombia     0.336194
Argentina    0.326113
Name: pais, dtype: float64
✅ 300 valores de país imputados

💾 cambio guardado en 'work_orders_clean.csv'


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - La limpieza y validación de los datos está desarrollada con criterio analítico y una justificación clara para cada transformación aplicada. Destaca el cuidado en la estandarización, manejo de duplicados e interpretación de inconsistencias financieras dentro del contexto del negocio.
</div>


In [19]:
# Imputación de valores faltantes - Variables categóricas y numéricas

# Definir columnas categóricas a imputar
columnas_categoricas = ['dispositivo', 'fuente_referencia', 'nombre_producto', 'categoria_producto']

# Imputar valores categóricos faltantes con 'unknown'
work_orders['dispositivo'] = work_orders['dispositivo'].fillna('unknown')
work_orders['fuente_referencia'] = work_orders['fuente_referencia'].fillna('unknown')
work_orders['nombre_producto'] = work_orders['nombre_producto'].fillna('unknown')
work_orders['categoria_producto'] = work_orders['categoria_producto'].fillna('unknown')

print("\n✅ VALORES FALTANTES CATEGÓRICOS DESPUÉS DE LA IMPUTACIÓN:")
for col in columnas_categoricas:
    faltantes_despues = work_orders[col].isnull().sum()
    print(f"   {col}: {faltantes_despues} valores faltantes")

# Imputar valores numéricos faltantes
work_orders['monto_descuento'] = work_orders['monto_descuento'].fillna(0)
work_orders['cantidad'] = work_orders['cantidad'].fillna(1)

print("\n✅ VALORES FALTANTES NUMÉRICOS DESPUÉS DE LA IMPUTACIÓN:")
for col in columnas_numericas:
    faltantes_despues = work_orders[col].isnull().sum()
    print(f"   {col}: {faltantes_despues} valores faltantes")

# Guardar cambios en dataset clean
work_orders.to_csv('work_orders_clean.csv', index=False)
print("\n💾 Dataset guardado en 'work_orders_clean.csv'")


✅ VALORES FALTANTES CATEGÓRICOS DESPUÉS DE LA IMPUTACIÓN:
   dispositivo: 0 valores faltantes
   fuente_referencia: 0 valores faltantes
   nombre_producto: 0 valores faltantes
   categoria_producto: 0 valores faltantes

✅ VALORES FALTANTES NUMÉRICOS DESPUÉS DE LA IMPUTACIÓN:
   monto_total: 0 valores faltantes
   precio_unitario: 50 valores faltantes
   cantidad: 0 valores faltantes
   monto_descuento: 0 valores faltantes

💾 Dataset guardado en 'work_orders_clean.csv'


In [20]:
#Imputar valores faltantes en precio_unitario 
def imputar_precios_con_datos_reales(df):
    """
    Imputa precio_unitario usando datos reales del cliente:
    precio_unitario = (monto_total + monto_descuento) / cantidad
    """
    print("🎯 IMPLEMENTANDO ESTRATEGIA DE IMPUTACIÓN INTELIGENTE")
    print("="*60)
    
    # Identificar casos con precio_unitario faltante
    mask_precio_faltante = df['precio_unitario'].isnull()
    casos_a_imputar = df[mask_precio_faltante].copy()
    
    print(f"📊 Casos a imputar: {len(casos_a_imputar)}")
    
    # Calcular precio_unitario usando datos reales
    precio_calculado = (casos_a_imputar['monto_total'] + casos_a_imputar['monto_descuento']) / casos_a_imputar['cantidad']
    
    # Aplicar la imputación
    df.loc[mask_precio_faltante, 'precio_unitario'] = precio_calculado
    
    # Mostrar resultados
    print(f"✅ {len(casos_a_imputar)} precios imputados con datos reales")
    print(f"📊 Precios faltantes restantes: {df['precio_unitario'].isnull().sum()}")
    
    return df

# Ejecutar estrategia
work_orders_imputado_real = imputar_precios_con_datos_reales(work_orders.copy())
work_orders = imputar_precios_con_datos_reales(work_orders)

# Guardar cambios en dataset clean
work_orders.to_csv('work_orders_clean.csv', index=False)
print("\n💾 Dataset guardado en 'work_orders_clean.csv'")

🎯 IMPLEMENTANDO ESTRATEGIA DE IMPUTACIÓN INTELIGENTE
📊 Casos a imputar: 50
✅ 50 precios imputados con datos reales
📊 Precios faltantes restantes: 0
🎯 IMPLEMENTANDO ESTRATEGIA DE IMPUTACIÓN INTELIGENTE
📊 Casos a imputar: 50
✅ 50 precios imputados con datos reales
📊 Precios faltantes restantes: 0

💾 Dataset guardado en 'work_orders_clean.csv'


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - La estrategia de imputación aplicada demuestra un enfoque consistente y bien pensado para preservar la calidad del dataset. Destaca especialmente el uso de distribuciones proporcionales y cálculos basados en datos reales para completar información faltante, ya que esto mantiene coherencia con el comportamiento original de los datos y fortalece la confiabilidad del análisis posterior.
</div>


In [21]:
#Validar consistencia de montos
work_orders['monto_calculado'] = work_orders['cantidad'] * work_orders['precio_unitario'] - work_orders['monto_descuento']
inconsistencias = abs(work_orders['monto_total'] - work_orders['monto_calculado']) > 0.01
print(f"⚠️ {inconsistencias.sum()} inconsistencias en montos detectadas")

# Ver ejemplos de inconsistencias
if inconsistencias.sum() > 0:
    print("\n📋 EJEMPLOS DE INCONSISTENCIAS:")
    ejemplos = work_orders[inconsistencias][['id_pedido', 'cantidad', 'precio_unitario',
                                           'monto_descuento', 'monto_total', 'monto_calculado']].head()
    print(ejemplos)
    
    # Calcular el impacto financiero
    diferencia_total = (work_orders.loc[inconsistencias, 'monto_total'] - 
                       work_orders.loc[inconsistencias, 'monto_calculado']).sum()
    print(f" Impacto financiero total: ${diferencia_total:.2f}")

⚠️ 1142 inconsistencias en montos detectadas

📋 EJEMPLOS DE INCONSISTENCIAS:
     id_pedido  cantidad  precio_unitario  monto_descuento  monto_total  \
2      order_2       2.0           102.99             10.0       195.99   
24    order_24       2.0           117.99              0.0       235.99   
35    order_35       2.0           467.34              5.0       929.69   
41    order_41       2.0            37.77             10.0        65.53   
136  order_136       2.0           211.05             15.0       407.09   

     monto_calculado  
2             195.98  
24            235.98  
35            929.68  
41             65.54  
136           407.10  
 Impacto financiero total: $0.48


In [22]:
# DETECCIÓN DE OUTLIERS USANDO MÉTODO IQR
def detectar_outliers_precios(df, columna='precio_unitario'):
    """
    Detecta outliers usando el método del Rango Intercuartílico (IQR)
    """
    Q1 = df[columna].quantile(0.25)
    Q3 = df[columna].quantile(0.75)
    IQR = Q3 - Q1
    
    # Límites para outliers
    limite_inferior = Q1 - 1.5 * IQR
    limite_superior = Q3 + 1.5 * IQR
    
    # Identificar outliers
    outliers_mask = (df[columna] < limite_inferior) | (df[columna] > limite_superior)
    outliers = df[outliers_mask]
    
    print(f"📊 RESULTADOS DETECCIÓN DE OUTLIERS:")
    print(f"   • Q1 (25%): ${Q1:.2f}")
    print(f"   • Q3 (75%): ${Q3:.2f}")
    print(f"   • IQR: ${IQR:.2f}")
    print(f"   • Límite inferior: ${limite_inferior:.2f}")
    print(f"   • Límite superior: ${limite_superior:.2f}")
    print(f"   • Outliers detectados: {len(outliers)}")
    
    return outliers, limite_inferior, limite_superior

# Ejecutar detección
outliers_detectados, limite_inf, limite_sup = detectar_outliers_precios(work_orders)

📊 RESULTADOS DETECCIÓN DE OUTLIERS:
   • Q1 (25%): $138.56
   • Q3 (75%): $380.78
   • IQR: $242.22
   • Límite inferior: $-224.77
   • Límite superior: $744.11
   • Outliers detectados: 6


In [23]:
# VERIFICACIÓN FINAL DE LA LIMPIEZA
print("\n" + "="*60)
print("🔍 VERIFICACIÓN FINAL DEL DATASET LIMPIO")
print("="*60)

# Verificar valores nulos restantes
print("\n📊 VALORES NULOS RESTANTES:")
nulos_finales = work_orders.isnull().sum()
for columna, cantidad in nulos_finales.items():
    if cantidad > 0:
        print(f"   ❌ {columna}: {cantidad} valores nulos")
    else:
        print(f"   ✅ {columna}: Sin valores nulos")

# Verificar duplicados
duplicados_finales = work_orders.duplicated().sum()
print(f"\n📊 DUPLICADOS RESTANTES: {duplicados_finales}")

# Verificar valores negativos en columnas numéricas
print("\n📊 VALORES NEGATIVOS RESTANTES:")
columnas_numericas = ['cantidad', 'precio_unitario', 'monto_descuento', 'monto_total']
for col in columnas_numericas:
    if col in work_orders.columns:
        negativos = (work_orders[col] < 0).sum()
        if negativos > 0:
            print(f"   ❌ {col}: {negativos} valores negativos")
        else:
            print(f"   ✅ {col}: Sin valores negativos")

# Resumen final
print("\n" + "="*60)
print("📈 RESUMEN FINAL DEL DATASET LIMPIO")
print("="*60)
print(f"📊 Total de registros: {len(work_orders):,}")
print(f"📊 Total de columnas: {len(work_orders.columns)}")
print(f"📊 Memoria utilizada: {work_orders.memory_usage(deep=True).sum() / 1024**2:.2f} MB")

# GUARDAR DATASET LIMPIO
print("\n" + "="*60)
print("💾 GUARDANDO DATASET LIMPIO")
print("="*60)

work_orders.to_csv('work_orders_clean.csv', index=False)
print("✅ Cambio guardado como 'work_orders_clean.csv'")
print("🎉 ¡LIMPIEZA DE DATOS COMPLETADA EXITOSAMENTE!")


🔍 VERIFICACIÓN FINAL DEL DATASET LIMPIO

📊 VALORES NULOS RESTANTES:
   ✅ id_pedido: Sin valores nulos
   ✅ id_usuario: Sin valores nulos
   ✅ fecha_pedido: Sin valores nulos
   ✅ pais: Sin valores nulos
   ✅ dispositivo: Sin valores nulos
   ✅ fuente_referencia: Sin valores nulos
   ✅ nombre_producto: Sin valores nulos
   ✅ categoria_producto: Sin valores nulos
   ✅ cantidad: Sin valores nulos
   ✅ precio_unitario: Sin valores nulos
   ✅ monto_descuento: Sin valores nulos
   ✅ monto_total: Sin valores nulos
   ✅ monto_calculado: Sin valores nulos

📊 DUPLICADOS RESTANTES: 0

📊 VALORES NEGATIVOS RESTANTES:
   ✅ cantidad: Sin valores negativos
   ✅ precio_unitario: Sin valores negativos
   ✅ monto_descuento: Sin valores negativos
   ✅ monto_total: Sin valores negativos

📈 RESUMEN FINAL DEL DATASET LIMPIO
📊 Total de registros: 25,000
📊 Total de columnas: 13
📊 Memoria utilizada: 12.38 MB

💾 GUARDANDO DATASET LIMPIO
✅ Cambio guardado como 'work_orders_clean.csv'
🎉 ¡LIMPIEZA DE DATOS COMPLET

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - La validación final del dataset refleja un proceso de limpieza completo y cuidadosamente documentado. La revisión de inconsistencias financieras, detección de outliers y verificación integral de nulos, duplicados y valores negativos aporta solidez al análisis posterior. Además, el cierre del proceso transmite un buen control sobre la calidad y confiabilidad de los datos preparados para negocio.
</div>


In [24]:
# 🚀 EJECUTAR PIPELINE PARA CATALOG

# Ejecutar pipeline completo para cada dataset
print("\n" + "="*60)
print("🔄 EJECUTANDO PIPELINE COMPLETO CATALOG")
print("="*60)

# Pipeline para Catalog  
resumen_catalog = pipeline_analisis_exploratorio(work_catalog, "Catalog")

print("🔍 ANÁLISIS INICIAL:")
print(f"Total datos faltantes:") 
print(f"{work_catalog.isna().sum()}")

print(f"Total datos únicos:")
print(f"categoria_producto: {work_catalog['categoria_producto'].nunique()}") 
print(f"nombre_producto: {work_catalog['nombre_producto'].nunique()}")
print(f"costo_unitario: {work_catalog['costo_unitario'].nunique()}")

print(f"Total datos duplicados:") 
print(f"{work_catalog.duplicated().value_counts()}")

# Ver distribuciones
print("\n📊 DISTRIBUCIONES:")
print(work_catalog['costo_unitario'].value_counts())  
print(work_catalog['proveedor'].value_counts())



🔄 EJECUTANDO PIPELINE COMPLETO CATALOG
📋 1. INFORMACIÓN GENERAL:
   • Filas: 7
   • Columnas: 4
🔍 ANÁLISIS INICIAL:
Total datos faltantes:
nombre_producto       0
categoria_producto    0
costo_unitario        0
proveedor             0
dtype: int64
Total datos únicos:
categoria_producto: 3
nombre_producto: 7
costo_unitario: 7
Total datos duplicados:
False    7
dtype: int64

📊 DISTRIBUCIONES:
189.31    1
280.68    1
16.60     1
25.21     1
176.64    1
17.21     1
10.12     1
Name: costo_unitario, dtype: int64
Fuller, Pena and Myers     1
Mcmillan-Rhodes            1
Greene-Smith               1
King Ltd                   1
Rivera, Carr and Finley    1
Bowers LLC                 1
Long-Reid                  1
Name: proveedor, dtype: int64


📊 REPORTE CATALOG - DATASET LIMPIO
✅ ESTADO: APROBADO - Sin necesidad de limpieza
✅ REGISTROS: 7 productos completos
✅ CALIDAD: 100% - Sin valores faltantes ni duplicados
✅ VALIDACIÓN: Todos los costos en rango válido
✅ CONCLUSIÓN: Listo para análisis de rentabilidad

In [25]:
# 🚀 EJECUTAR PIPELINE COMPLETO PARA MARKETING
# Ejecutar pipeline completo para cada dataset
print("\n" + "="*60)
print("🔄 EJECUTANDO PIPELINE COMPLETO MARKETING")
print("="*60)

# Pipeline para Marketing
resumen_marketing = pipeline_analisis_exploratorio(work_marketing, "Marketing")

print("🔍 ANÁLISIS INICIAL:")
print(f"Total datos faltantes:") 
print(f"{work_marketing.isna().sum()}")

print(f"Total de datos únicos:")
print(f"pais: {work_marketing['pais'].nunique()}")
print(f"canal: {work_marketing['canal'].nunique()}")
print(f"id_campaña: {work_marketing['id_campaña'].nunique()}")

print(f"Total datos duplicados:") 
print(f"{work_marketing.duplicated().value_counts()}")

# Ver distribuciones
print("\n📊 DISTRIBUCIONES:")
print(work_marketing['gasto'].value_counts())


🔄 EJECUTANDO PIPELINE COMPLETO MARKETING
📋 1. INFORMACIÓN GENERAL:
   • Filas: 1,620
   • Columnas: 5
🔍 ANÁLISIS INICIAL:
Total datos faltantes:
fecha           0
pais            0
id_campaña      0
canal         101
gasto           0
dtype: int64
Total de datos únicos:
pais: 3
canal: 3
id_campaña: 9
Total datos duplicados:
False    1620
dtype: int64

📊 DISTRIBUCIONES:
619.11     2
2696.62    2
1018.16    2
1731.35    2
2775.27    2
          ..
1651.40    1
1545.07    1
1917.53    1
723.60     1
2606.24    1
Name: gasto, Length: 1612, dtype: int64


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - La revisión de los datasets complementarios mantiene una estructura clara y consistente con el análisis principal. La validación de calidad en catalog y marketing permite confirmar rápidamente el estado de los datos y detectar posibles puntos de atención antes de avanzar al análisis de negocio. Destaca también la organización del pipeline y la claridad en la interpretación de los resultados obtenidos.
</div>


Proceder con la limpieza en problemas identificados en Marketing:
🔴 CRÍTICOS:
   • Canal: 101 valores faltantes (6.2% del dataset)
   • Fecha: Tipo incorrecto (object en lugar de datetime)
ESTRATEGIAS DE LIMPIEZA A IMPLEMENTAR:
1. CONVERSIÓN DE TIPOS DE DATOS:
   ✅ Convertir 'fecha' de object a datetime64
   ✅ Renombrar columna a 'fecha_campaña' para mayor claridad
2️. IMPUTACIÓN DE VALORES FALTANTES:
   ✅ Canal faltante (101 registros):
      • Estrategia: Inferir canal desde imputación inteligente'id_campaña'

In [26]:
# Convertir columna de work_marketing (fecha) a tipo fecha
work_marketing['fecha'] = pd.to_datetime(work_marketing["fecha"], errors="coerce")

# Cambiar nombre de columna usando rename()
work_marketing = work_marketing.rename(columns={'fecha': 'fecha_campaña'})

#Verificar cambios 
work_marketing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1620 entries, 0 to 1619
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   fecha_campaña  1620 non-null   datetime64[ns]
 1   pais           1620 non-null   object        
 2   id_campaña     1620 non-null   object        
 3   canal          1519 non-null   object        
 4   gasto          1620 non-null   float64       
dtypes: datetime64[ns](1), float64(1), object(3)
memory usage: 63.4+ KB


In [27]:
# 🔧 LIMPIEZA COMPLETA DEL DATASET MARKETING
print("="*60)
print("🔧 IMPUTACIÓN INTELIGENTE DE CANAL DESDE ID_CAMPAÑA")
print("="*60)

# 📊 MOSTRAR ESTADO ANTES DE LA LIMPIEZA
print("\n📊 ESTADO ANTES DE LA LIMPIEZA:")
print(f"   • Valores faltantes en canal: {work_marketing['canal'].isnull().sum()}")
print(f"   • Canales únicos: {work_marketing['canal'].nunique()}")

# 🔍 ANALIZAR PATRÓN DE ID_CAMPAÑA
print("\n🔍 ANALIZANDO PATRÓN DE ID_CAMPAÑA:")
print("Ejemplos de id_campaña:")
print(work_marketing['id_campaña'].head(10).tolist())

# 🧠 FUNCIÓN PARA EXTRAER CANAL DESDE ID_CAMPAÑA
def extraer_canal_desde_id(id_campaña):
    """
    Extrae el canal desde el id_campaña
    Patrón esperado: [canal]_[país]
    """
    if pd.isna(id_campaña):
        return 'unknown'

    # Dividir por '_' y tomar la primera parte
    partes = str(id_campaña).split('_')
    if len(partes) >= 2:
        canal_extraido = partes[0].lower()

        # Mapear canales conocidos
        mapeo_canales = {
            'organic': 'organic',
            'paid_search': 'paid_search', 
            'social': 'social',
            'paidsearch': 'paid_search',  # variaciones posibles
            'ppc': 'paid_search',
            'sem': 'paid_search'
        }

        return mapeo_canales.get(canal_extraido, 'unknown')
    else:
        return 'unknown'

# 📋 IDENTIFICAR REGISTROS CON CANAL FALTANTE
mask_canal_faltante = work_marketing['canal'].isnull()
registros_faltantes = work_marketing[mask_canal_faltante].copy()

print(f"\n📋 REGISTROS CON CANAL FALTANTE: {len(registros_faltantes)}")
print("Ejemplos de id_campaña sin canal:")
print(registros_faltantes[['id_campaña', 'canal']].head())

# 🔧 APLICAR IMPUTACIÓN INTELIGENTE
print("\n🔧 APLICANDO IMPUTACIÓN INTELIGENTE...")

# Crear columna temporal con canal extraído
work_marketing['canal_extraido'] = work_marketing['id_campaña'].apply(extraer_canal_desde_id)

# Imputar solo donde canal es nulo
work_marketing.loc[mask_canal_faltante, 'canal'] = work_marketing.loc[mask_canal_faltante, 'canal_extraido']

# 📊 MOSTRAR RESULTADOS
print("\n📊 RESULTADOS DE LA IMPUTACIÓN:")
print(f"   • Valores imputados: {mask_canal_faltante.sum()}")
print(f"   • Valores faltantes restantes: {work_marketing['canal'].isnull().sum()}")

🔧 IMPUTACIÓN INTELIGENTE DE CANAL DESDE ID_CAMPAÑA

📊 ESTADO ANTES DE LA LIMPIEZA:
   • Valores faltantes en canal: 101
   • Canales únicos: 3

🔍 ANALIZANDO PATRÓN DE ID_CAMPAÑA:
Ejemplos de id_campaña:
['organic_Mexico', 'paid_search_Mexico', 'social_Mexico', 'organic_Colombia', 'paid_search_Colombia', 'social_Colombia', 'organic_Argentina', 'paid_search_Argentina', 'social_Argentina', 'organic_Mexico']

📋 REGISTROS CON CANAL FALTANTE: 101
Ejemplos de id_campaña sin canal:
             id_campaña canal
98     social_Argentina   NaN
99       organic_Mexico   NaN
100  paid_search_Mexico   NaN
101       social_Mexico   NaN
102    organic_Colombia   NaN

🔧 APLICANDO IMPUTACIÓN INTELIGENTE...

📊 RESULTADOS DE LA IMPUTACIÓN:
   • Valores imputados: 101
   • Valores faltantes restantes: 0


In [28]:
# VALIDAR CONSISTENCIA DE DATOS
print("\n🔍 VALIDACIONES ADICIONALES:")
# Verificar que no hay gastos negativos o cero
gastos_negativos = (work_marketing['gasto'] <= 0).sum()
print(f"   • Gastos negativos o cero: {gastos_negativos}")

# 📊 MOSTRAR ESTADO DESPUÉS DE LA LIMPIEZA
print("\n📊 ESTADO DESPUÉS DE LA LIMPIEZA:")
print(f"   • Total registros: {len(work_marketing):,}")
print(f"   • Valores faltantes en canal: {work_marketing['canal'].isnull().sum()}")
print(f"   • Tipo de fecha: {work_marketing['fecha_campaña'].dtype}")
print(f"   • Canales únicos: {work_marketing['canal'].nunique()}")
print(f"   • Distribución de canales:")
print(work_marketing['canal'].value_counts())

# 4️⃣ GUARDAR DATASET LIMPIO
work_marketing.to_csv('marketing_clean.csv', index=False)
print("\n💾 Dataset limpio guardado como 'marketing_clean.csv'")

# 5️⃣ VERIFICACIÓN FINAL
print("\n🎉 VERIFICACIÓN FINAL:")
print("✅ Sin valores faltantes en columnas críticas")
print("✅ Fechas en formato datetime correcto")
print("✅ Canales identificados para análisis de ROI")
print("✅ Dataset listo para análisis de marketing")


🔍 VALIDACIONES ADICIONALES:
   • Gastos negativos o cero: 0

📊 ESTADO DESPUÉS DE LA LIMPIEZA:
   • Total registros: 1,620
   • Valores faltantes en canal: 0
   • Tipo de fecha: datetime64[ns]
   • Canales únicos: 4
   • Distribución de canales:
social         540
organic        540
paid_search    507
unknown         33
Name: canal, dtype: int64

💾 Dataset limpio guardado como 'marketing_clean.csv'

🎉 VERIFICACIÓN FINAL:
✅ Sin valores faltantes en columnas críticas
✅ Fechas en formato datetime correcto
✅ Canales identificados para análisis de ROI
✅ Dataset listo para análisis de marketing


---
**📦 Exportación**: Una vez finalizada la limpieza, se exportan los datasets para utilizarlos en la última etapa del proyecto.

In [29]:
# Exportar los datasets limpios 
work_orders.to_csv('orders_clean.csv', index=False)
work_catalog.to_csv('catalog_clean.csv', index=False) 
work_marketing.to_csv('marketing_clean.csv', index=False)

print("✅ Datasets exportados correctamente:")
print("- orders_clean.csv (datos de pedidos limpios)")
print("- catalog_clean.csv (costos de productos)")
print("- marketing_clean.csv (gastos de marketing)")

✅ Datasets exportados correctamente:
- orders_clean.csv (datos de pedidos limpios)
- catalog_clean.csv (costos de productos)
- marketing_clean.csv (gastos de marketing)


In [30]:
# ✅ CORRECTO - Verificar cada archivo individualmente
import os
archivos_a_verificar = [
    'work_orders_clean.csv', 
    'catalog_clean.csv', 
    'marketing_clean.csv'
]

print("🔍 VERIFICANDO ARCHIVOS EXPORTADOS:")
print("="*50)

# todos_existen = True
for archivo in archivos_a_verificar:
    if os.path.exists(archivo):
        size = os.path.getsize(archivo) / 1024  # KB
        print(f"✅ {archivo}: {size:.1f} KB")
    else:
        print(f"❌ {archivo}: No encontrado")
        todos_existen = False


🔍 VERIFICANDO ARCHIVOS EXPORTADOS:
✅ work_orders_clean.csv: 2714.3 KB
✅ catalog_clean.csv: 0.4 KB
✅ marketing_clean.csv: 97.6 KB


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - La imputación inteligente aplicada al dataset de marketing demuestra un razonamiento sólido al aprovechar la estructura de los identificadores de campaña para recuperar información faltante. La validación posterior confirma un proceso de limpieza consistente y preparado para el análisis de ROI.
</div>


---

## 🔹 Paso 2: Analizar si el negocio es rentable

### 2.1 Cálculo de KPIs principales

**🎯 Objetivo:** Calcular los indicadores clave del negocio para evaluar ingresos, costos y rentabilidad.


Se usan los 3 datasets (`orders`, `catalog`, `marketing_spend`):

**📊 Parte 1: Rentabilidad del negocio**
- ¿Cuál es el ingreso total (revenue)? 
- ¿Cuál es el costo total? 
- ¿Cuánto se ha invertido en marketing? 
- ¿El negocio es rentable? (calcular profit)  

---

**📈 Parte 2: Comportamiento de ventas**
- ¿Cuál es el ticket promedio por orden? 
- ¿Cuál es la cantidad promedio de productos por orden? 
- ¿Cuál es el producto más vendido?
- ¿Cuánto se ha gastado en marketing por canal? 


In [31]:
# Cargar los datasets LIMPIOS para el análisis de KPIs
orders_clean = pd.read_csv('work_orders_clean.csv')
catalog_clean = pd.read_csv('catalog_clean.csv') 
marketing_clean = pd.read_csv('marketing_clean.csv')

In [32]:
# Revenue Total (Ingresos)
# REVENUE TOTAL - Suma de todos los ingresos
revenue_total = orders_clean['monto_total'].sum()

print("REVENUE (INGRESOS)")
print(f"Revenue total: ${revenue_total:,.2f}")

REVENUE (INGRESOS)
Revenue total: $51,989,756.86


In [33]:
# COSTOS TOTALES - Unir orders con catalog para obtener costos
orders_with_costs = orders_clean.merge(
    catalog_clean[['nombre_producto', 'costo_unitario']], 
    on='nombre_producto', 
    how='left'
)

# Calcular costo total por pedido
orders_with_costs['costo_total'] = orders_with_costs['cantidad'] * orders_with_costs['costo_unitario']

# Sumar todos los costos
costos_totales = orders_with_costs['costo_total'].sum()

print("\n💸 COSTOS DE PRODUCTOS")
print(f"Costos totales: ${costos_totales:,.2f}")


💸 COSTOS DE PRODUCTOS
Costos totales: $43,128,248.35


In [34]:
# GASTO EN MARKETING - Suma de todos los gastos de marketing
gasto_marketing = marketing_clean['gasto'].sum()

print("\n📢 GASTO EN MARKETING")
print(f"Gasto total en marketing: ${gasto_marketing:,.2f}")
print(f"Número de campañas: {len(marketing_clean):,}")


📢 GASTO EN MARKETING
Gasto total en marketing: $2,871,843.53
Número de campañas: 1,620


In [35]:
# PROFIT - Revenue menos todos los costos
profit = revenue_total - costos_totales - gasto_marketing

print("\n📈 PROFIT (GANANCIA)")
print(f"Revenue: ${revenue_total:,.2f}")
print(f"- Costos productos: ${costos_totales:,.2f}")
print(f"- Gasto marketing: ${gasto_marketing:,.2f}")
print("=" * 40)
print(f"PROFIT FINAL: ${profit:,.2f}")

# Análisis de rentabilidad
if profit > 0:
    print("✅ El negocio ES RENTABLE")
    margen = (profit / revenue_total) * 100
    print(f"Margen de ganancia: {margen:.2f}%")
else:
    print("❌ El negocio NO es rentable")


📈 PROFIT (GANANCIA)
Revenue: $51,989,756.86
- Costos productos: $43,128,248.35
- Gasto marketing: $2,871,843.53
PROFIT FINAL: $5,989,664.98
✅ El negocio ES RENTABLE
Margen de ganancia: 11.52%


In [36]:
# Calcular ticket promedio por orden
ticket_promedio = orders_clean['monto_total'].mean()
print("🎫 TICKET PROMEDIO POR ORDEN")
print(f"Ticket promedio: ${ticket_promedio:,.2f}")
print(f"Total órdenes: {len(orders_clean):,}")

🎫 TICKET PROMEDIO POR ORDEN
Ticket promedio: $2,079.59
Total órdenes: 25,000


In [37]:
# Cantidad promedio de productos por orden
cantidad_promedio = orders_clean['cantidad'].mean()
print("📦 CANTIDAD PROMEDIO DE PRODUCTOS POR ORDEN")
print(f"Cantidad promedio: {cantidad_promedio:.2f} productos por orden")

# Análisis adicional
print(f"Cantidad mínima: {orders_clean['cantidad'].min()}")
print(f"Cantidad máxima: {orders_clean['cantidad'].max()}")
print(f"Mediana: {orders_clean['cantidad'].median()}")

📦 CANTIDAD PROMEDIO DE PRODUCTOS POR ORDEN
Cantidad promedio: 7.10 productos por orden
Cantidad mínima: 1.0
Cantidad máxima: 20000.0
Mediana: 2.0


In [38]:
# Producto más vendido (por cantidad total vendida)
productos_vendidos = orders_clean.groupby('nombre_producto')['cantidad'].sum().sort_values(ascending=False)
print("\n🏆 PRODUCTOS MÁS VENDIDOS (por cantidad)")
print(productos_vendidos.head())

# También puedes analizar por número de órdenes
productos_por_ordenes = orders_clean['nombre_producto'].value_counts()
print("\n📊 PRODUCTOS MÁS VENDIDOS (por número de órdenes)")
print(productos_por_ordenes.head())


🏆 PRODUCTOS MÁS VENDIDOS (por cantidad)
nombre_producto
Laptop-Gaming-16GB    144202.0
Vacuum-Pro-Black        6290.0
Blender-XL-Red          6282.0
Jacket-Winter-M         6266.0
Sneakers-Urban-42       6191.0
Name: cantidad, dtype: float64

📊 PRODUCTOS MÁS VENDIDOS (por número de órdenes)
Blender-XL-Red        4179
Vacuum-Pro-Black      4176
Jacket-Winter-M       4176
Sneakers-Urban-42     4148
Laptop-Gaming-16GB    2782
Name: nombre_producto, dtype: int64


In [39]:
# Gasto en marketing por canal
gasto_por_canal = marketing_clean.groupby('canal')['gasto'].sum().sort_values(ascending=False)
print("\n💰 GASTO EN MARKETING POR CANAL")
print(gasto_por_canal)

# Porcentaje del gasto total por canal
porcentaje_por_canal = (gasto_por_canal / gasto_por_canal.sum() * 100).round(2)
print("\n📊 PORCENTAJE DEL GASTO POR CANAL")
print(porcentaje_por_canal)


💰 GASTO EN MARKETING POR CANAL
canal
social         976818.37
organic        972650.96
paid_search    863088.21
unknown         59285.99
Name: gasto, dtype: float64

📊 PORCENTAJE DEL GASTO POR CANAL
canal
social         34.01
organic        33.87
paid_search    30.05
unknown         2.06
Name: gasto, dtype: float64


In [40]:
# Ingresos por país
ingresos_por_pais = orders_clean.groupby('pais')['monto_total'].sum().sort_values(ascending=False)

print("INGRESOS TOTALES POR PAÍS:")
print("="*40)
for pais, ingreso in ingresos_por_pais.items():
    print(f"{pais}: ${ingreso:,.2f}")

# Calcular porcentajes
total_ingresos = ingresos_por_pais.sum()
porcentajes = (ingresos_por_pais / total_ingresos * 100).round(2)

print("\n📊 PORCENTAJE DE INGRESOS POR PAÍS:")
print("="*40)
for pais, porcentaje in porcentajes.items():
    print(f"{pais}: {porcentaje}%")

INGRESOS TOTALES POR PAÍS:
Argentina: $20,768,087.15
Mexico: $19,805,737.65
Colombia: $11,415,932.06

📊 PORCENTAJE DE INGRESOS POR PAÍS:
Argentina: 39.95%
Mexico: 38.1%
Colombia: 21.96%


In [41]:
# Análisis completo por país
analisis_pais = work_orders.groupby('pais').agg({
    'monto_total': ['sum', 'mean', 'count'],
    'cantidad': 'sum'
}).round(2)

analisis_pais.columns = ['Revenue_Total', 'Ticket_Promedio', 'Num_Ordenes', 'Productos_Vendidos']
print("\n📈 ANÁLISIS DETALLADO POR PAÍS:")
print(analisis_pais)


📈 ANÁLISIS DETALLADO POR PAÍS:
           Revenue_Total  Ticket_Promedio  Num_Ordenes  Productos_Vendidos
pais                                                                      
Argentina    20768087.15          2547.61         8152             62294.0
Colombia     11415932.06          1359.53         8397             42573.0
Mexico       19805737.65          2343.60         8451             72715.0


Análisis de los resultados generados por RappiPlus:

Cálculo de KPIs principales 🎯
Parte 1 - Rentabilidad del negocio: Revenue 51,989,756.86 USD; Costo total 43,128,248.35 USD; Marketing 2,871,843.53 USD; Profit 5,989,664.98 USD; Margen neto 11.52% (ROAS ≈ 18x).ROAS = Return on Advertising Spend (Retorno de la Inversión Publicitaria) Es una métrica que mide cuántos dólares de ingresos generas por cada dólar invertido en marketing.

Parte 2 - Comportamiento de ventas: Ticket promedio 2,079.59 USD; Productos por orden 7.10; Producto más vendido: Laptop-Gaming-16GB (144,202 unidades en 2,782 órdenes); Gasto en marketing por canal: Social 34.01%, Organic 33.87%, Paid Search 30.05%, Unknown 2.06%. 

¿El negocio realmente está generando ganancias?

Sí. El negocio es rentable y muestra una operación financieramente sólida.
Ingresos totales: 51.99M USD
Ganancia neta: 5.99M USD
Margen neto: 11.52%
ROAS: 18x

Hallazgos clave
✅ Por cada 1 USD invertido en marketing, la empresa genera 18 USD en ingresos.
✅ El negocio no solo vende alto volumen; también convierte ventas en utilidad real.
✅ El ticket promedio de 2,079 USD y 7.1 productos por orden reflejan clientes de alto valor.
⚠️ La Laptop-Gaming-16GB concentra gran parte de las ventas, generando dependencia de un producto estrella.
✅ El gasto en marketing está balanceado entre Social, Orgánico y Paid Search, reduciendo riesgo comercial.
✅ El negocio ya es rentable, pero el nivel de costos sigue siendo elevado, ya que más del 80% de los ingresos se absorben en costos operativos. Esto abre una oportunidad estratégica en: optimización logística, negociación con proveedores, eficiencia operativa, y mejora del mix de productos.
✅ La adquisición de clientes no depende de un solo canal.
✅ Existe diversificación de tráfico y captación.
✅ El canal orgánico tiene peso importante, lo cual reduce dependencia de publicidad pagada.
✅ El 2.06% “Unknown” sugiere una oportunidad de mejorar de trazabilidad y atribución de campañas.
✅ Argentina tiene el mercado más rentable, México muy cerca de Argentina y Colombia cuenta con el mercado más pequeño. 

Conclusión ejecutiva
RappiPlus demuestra que tiene capacidad de crecimiento y rentabilidad.
El siguiente reto estratégico es mejorar eficiencia operativa, optimizar costos y escalar los productos y canales más rentables.




## 🔹 Paso 3: Entender dónde se pierden los usuarios (funnel de conversión)

**🎯 Objetivo:** Analizar el comportamiento de los usuarios para identificar en qué etapa del proceso se pierden.


⚙️**Conexión a la base de datos**:  
Se ejecuta la línea de configuración para conectar con la base de datos y aplicar consultas SQL en la tabla **events**.

---

**📊 Parte 1: Construcción del funnel**
- ¿Cuántos usuarios llegan a cada etapa del funnel?  
- Se calcula el número de usuarios únicos por `nombre_evento`  
- Se ordenan los eventos según el flujo del usuario  

---

**📉 Parte 2: Análisis de conversión**
- Se calcula la tasa de conversión entre cada paso del funnel  
- Se identifica en qué etapa se pierde la mayor cantidad de usuarios  
- ¿Cuál es la tasa de conversión final?


---

In [42]:
import pandas as pd
from sqlalchemy import create_engine

# ======================
# Conexión (NO modificar)
# ======================
db_config = {
    'user': 'practicum_student',
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7',
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
    'port': 5432,
    'db': 'data-analyst-production-db-en'
}

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Atención</b> ⚠️ - La celda contiene credenciales de base de datos (usuario y contraseña) hardcodeadas como cadenas de texto. Si bien marca estas líneas como "(NO modificar)" para este ejercicio, es importante que sepas que en proyectos reales esta es una práctica de seguridad crítica a evitar.

Propuesta:
- En proyectos reales, almacena las credenciales en variables de entorno (.env) o en secretos del sistema
- Usa librerías como python-dotenv para cargar credenciales desde un archivo .env que NO está en control de versiones
- Nunca hagas push de credenciales a repositorios (GitHub, etc.)

Aunque en TripleTen las credenciales se proporcionan para que puedas trabajar directamente en las actividades, es importante que conozcas esta buena práctica desde ahora. Comprender cómo manejar credenciales de forma segura será muy útil cuando trabajes en entornos de producción reales.

</div>

In [43]:

# Explorar tabla events
# =========================
query_events = '''
SELECT *
FROM events;
'''
events = pd.read_sql(query_events, con=engine)
events.head()


,id_usuario,id_sesion,nombre_evento,timestamp_evento,pais,dispositivo,fuente_referencia,categoria_producto
0,user_6772,6a97f2af-32ae-4186-8c92-04025be1a27b,first_visit,2025-05-17,Colombia,desktop,organic,Moda
1,user_5883,369b767c-1c33-4b2f-a652-c7c0ef92cfc9,add_to_cart,2025-02-23,Mexico,mobile,social,Hogar
2,user_5946,60039041-e78b-474c-87b3-c0b7e9c30708,add_payment_info,2025-05-15,Colombia,desktop,social,Electronica
3,user_827,18252a64-f389-4ef7-9e58-dadad4a3491e,purchase,2025-03-31,Mexico,mobile,social,Moda
4,user_2361,221b364e-cdc5-4668-b698-18d5ba849a67,first_visit,2025-01-22,Argentina,desktop,paid_search,Electronica


In [44]:
# PARTE 1: Totales del funnel
# ======================
# Eventos disponibles
query_eventos_disponibles = '''
SELECT 
    nombre_evento,
    COUNT(DISTINCT id_usuario) as usuarios_unicos
FROM events
GROUP BY nombre_evento
ORDER BY usuarios_unicos DESC;
'''
eventos_disponibles = pd.read_sql(query_eventos_disponibles, con=engine)
eventos_disponibles

,nombre_evento,usuarios_unicos
0,first_visit,7796
1,add_to_cart,7634
2,select_item,7582
3,begin_checkout,7208
4,add_payment_info,6250
5,purchase,6240


In [45]:
# PARTE 2: Conversiones
# ======================

query_funnel_completo = '''
WITH funnel_analysis AS (
    SELECT
        'first_visit' as etapa,
        1 as orden_etapa,
        COUNT(DISTINCT id_usuario) as usuarios_unicos
    FROM events
    WHERE nombre_evento = 'first_visit'
    
    UNION ALL
    
    SELECT
        'select_item' as etapa,
        2 as orden_etapa,
        COUNT(DISTINCT id_usuario) as usuarios_unicos
    FROM events
    WHERE nombre_evento = 'select_item'
    
    UNION ALL
    
    SELECT
        'add_to_cart' as etapa,
        3 as orden_etapa,
        COUNT(DISTINCT id_usuario) as usuarios_unicos
    FROM events
    WHERE nombre_evento = 'add_to_cart'
    
    UNION ALL
    
    SELECT
        'begin_checkout' as etapa,
        4 as orden_etapa,
        COUNT(DISTINCT id_usuario) as usuarios_unicos
    FROM events
    WHERE nombre_evento = 'begin_checkout'
    
    UNION ALL
    
    SELECT
        'add_payment_info' as etapa,
        5 as orden_etapa,
        COUNT(DISTINCT id_usuario) as usuarios_unicos
    FROM events
    WHERE nombre_evento = 'add_payment_info'
    
    UNION ALL
    
    SELECT
        'purchase' as etapa,
        6 as orden_etapa,
        COUNT(DISTINCT id_usuario) as usuarios_unicos
    FROM events
    WHERE nombre_evento = 'purchase'
),
funnel_with_conversion AS (
    SELECT
        etapa,
        orden_etapa,
        usuarios_unicos,
        LAG(usuarios_unicos) OVER (ORDER BY orden_etapa) as usuarios_etapa_anterior,
        CASE 
            WHEN LAG(usuarios_unicos) OVER (ORDER BY orden_etapa) IS NULL THEN 100.0
            ELSE ROUND(
                (usuarios_unicos * 100.0) / LAG(usuarios_unicos) OVER (ORDER BY orden_etapa), 
                2
            )
        END as tasa_conversion,
        CASE 
            WHEN LAG(usuarios_unicos) OVER (ORDER BY orden_etapa) IS NULL THEN 0
            ELSE LAG(usuarios_unicos) OVER (ORDER BY orden_etapa) - usuarios_unicos
        END as usuarios_perdidos,
        ROUND(
            (usuarios_unicos * 100.0) / FIRST_VALUE(usuarios_unicos) OVER (ORDER BY orden_etapa), 
            2
        ) as conversion_desde_inicio
    FROM funnel_analysis
)
SELECT
    etapa,
    usuarios_unicos,
    usuarios_etapa_anterior,
    tasa_conversion,
    usuarios_perdidos,
    conversion_desde_inicio
FROM funnel_with_conversion
ORDER BY orden_etapa;
'''
# Ejecutar la consulta usando pandas
funnel_completo = pd.read_sql(query_funnel_completo, con=engine)

# Crear formato personalizado - VERSIÓN CORREGIDA
print("ANÁLISIS DE CONVERSIÓN DEL FUNNEL")
print("=" * 60)
print(f"{'ETAPA':<18} {'USUARIOS':<10} {'CONVERSIÓN':<12} {'PERDIDOS':<10}")
print("-" * 60)

for _, row in funnel_completo.iterrows():
    etapa = str(row['etapa'])  # ✅ Convertir a string
    usuarios = f"{int(row['usuarios_unicos']):,}"
    conversion = f"{float(row['tasa_conversion']):.2f}%"
    perdidos = f"{int(row['usuarios_perdidos']):,}"
    
    print(f"{etapa:<18} {usuarios:<10} {conversion:<12} {perdidos:<10}")

ANÁLISIS DE CONVERSIÓN DEL FUNNEL
ETAPA              USUARIOS   CONVERSIÓN   PERDIDOS  
------------------------------------------------------------
first_visit        7,796      100.00%      0         
select_item        7,582      97.26%       214       
add_to_cart        7,634      100.69%      -52       
begin_checkout     7,208      94.42%       426       
add_payment_info   6,250      86.71%       958       
purchase           6,240      99.84%       10        


In [46]:
# En tu notebook, después del análisis del funnel
funnel_data = pd.DataFrame({
    'etapa': ['first_visit', 'select_item', 'add_to_cart', 'begin_checkout', 'add_payment_info', 'purchase'],
    'usuarios_unicos': [7796, 7582, 7634, 7208, 6250, 6240],
    'tasa_conversion': [100.00, 97.26, 100.69, 94.42, 86.71, 99.84],
    'usuarios_perdidos': [0, 214, -52, 426, 958, 10]
})

# Exportar para Power BI
funnel_data.to_csv('funnel_analysis_clean.csv', index=False)

Análisis Ejecutivo del Funnel de Conversión

🎯 ¿Dónde se pierden los usuarios?. El funnel muestra una conversión sólida: de 7,796 usuarios, 6,240 completan la compra (~80% de conversión final).
El principal punto de fuga ocurre entre: begin_checkout → **add_payment_info** (Mayor DROP-OFF). Pérdida: 13.3% de usuarios (958).

⚠️ El mayor problema está en add_payment_info. La caída ocurre antes del pago, lo que sugiere dificultad en: formularios, costos inesperados, métodos de pago, o confianza del usuario.
✅ El sitio convierte bien desde el inicio. El 97% de los usuarios interactúa con productos, señal de tráfico calificado y buena experiencia inicial.
✅ La intención de compra es alta. Una vez que el usuario agrega datos de pago, casi todos finalizan la compra (99.8% de conversión).
✅ Existe una señal positiva de recompra o compra rápida. Hay más usuarios en add_to_cart que en select_item, indicando usuarios recurrentes o procesos de compra eficientes.

📈 Impacto de negocio
El negocio no necesita más tráfico; necesita perder menos usuarios en add_payment_info.
Reducir la fuga del 13.3% tendría impacto directo en ventas sin aumentar inversión en marketing.

🏁 Conclusión:
El funnel es saludable y la intención de compra es fuerte.
La principal oportunidad de crecimiento está en optimizar el add_payment_info para convertir más usuarios que ya tienen intención de compra.  
En el contexto de RappiPlus: Como es un servicio de suscripción, los usuarios podrían abandonar porque: No entienden claramente qué incluye la suscripción, Los beneficios no justifican el costo o Prefieren pagar por pedido individual. Aquí están diferentes accionables para iterar mejoras y experiencias con el cliente para reducir su abandono. 
El negocio ya logra atraer y convencer al usuario. El reto ahora es eliminar fricción antes del pago.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - El cálculo de KPIs y la interpretación de resultados muestran una conexión clara entre métricas técnicas y decisiones de negocio. La lectura ejecutiva de rentabilidad y comportamiento comercial aporta valor analítico al proyecto.
</div>


---

## 🔹 Paso 4: Evaluar si los usuarios regresan (retención por cohortes)

**🎯 Objetivo:** Analizar la retención de usuarios para entender si regresan después de registrarse.

**Tablas**

- `users` 
- `user_activity` 

---

1. Se identifica la cohorte de cada usuario según el **mes de registro**.


2. Se calcula la retención semanal: cuántos usuarios **se mantienen activos** en cada semana desde su registro.
   - `retenido_w1`: usuarios activos en la semana 1  
   - `retenido_w2`: usuarios activos en la semana 2  
   - `retenido_w3`: usuarios activos en la semana 3  


3. Se calcula el porcentaje de retención para cada semana, dividiendo los usuarios retenidos entre los clientes iniciales de la cohorte:  
   - `semana_1`: porcentaje de usuarios retenidos en la semana 1  
   - `semana_2`: porcentaje de usuarios retenidos en la semana 2  
   - `semana_3`: porcentaje de usuarios retenidos en la semana 3

Se revisa que la columna de fecha esté en formato correcto (`DATE`).  
Se realiza la conversión usando: `CAST(fecha_registro AS DATE)`

In [47]:
# Explorar tabla users
# =========================
query_users = '''
SELECT *
FROM users;
'''
users = pd.read_sql(query_users, con=engine)
users.info()
users.head(3)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   id_usuario      8000 non-null   object
 1   fecha_registro  8000 non-null   object
 2   país            8000 non-null   object
 3   dispositivo     8000 non-null   object
 4   tipo_plan       8000 non-null   object
dtypes: object(5)
memory usage: 312.6+ KB


,id_usuario,fecha_registro,país,dispositivo,tipo_plan
0,user_0,2025-01-29,Mexico,mobile,free
1,user_1,2025-01-07,Mexico,mobile,free
2,user_2,2025-03-12,Argentina,mobile,free


In [48]:
# Cambio en users (Dtype fecha_registro a DATE)
query_users_converted = '''
SELECT 
    id_usuario,
    CAST(fecha_registro AS DATE) as fecha_registro,
    país,
    dispositivo,
    tipo_plan
FROM users;
'''
users = pd.read_sql(query_users_converted, con=engine)

# Conversión explícita en pandas
users['fecha_registro'] = pd.to_datetime(users['fecha_registro'])

# Verificar el cambio
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   id_usuario      8000 non-null   object        
 1   fecha_registro  8000 non-null   datetime64[ns]
 2   país            8000 non-null   object        
 3   dispositivo     8000 non-null   object        
 4   tipo_plan       8000 non-null   object        
dtypes: datetime64[ns](1), object(4)
memory usage: 312.6+ KB


In [49]:
# Explorar tabla user_activity
# =========================
query_user_activity = '''
SELECT *
FROM user_activity
LIMIT 5; 
'''
user_activity = pd.read_sql(query_user_activity, con=engine)
user_activity.info()
user_activity.head(3)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 4 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   id_usuario             5 non-null      object
 1   fecha_actividad        5 non-null      object
 2   dias_despues_registro  5 non-null      int64 
 3   activo                 5 non-null      int64 
dtypes: int64(2), object(2)
memory usage: 288.0+ bytes


,id_usuario,fecha_actividad,dias_despues_registro,activo
0,user_0,2025-02-05,7,0
1,user_0,2025-02-12,14,1
2,user_0,2025-02-19,21,1


In [50]:
#Cambio en user_activity (Dtype fecha_actividad a DATE)
query_user_activity_converted = '''
SELECT
    id_usuario,
    CAST(fecha_actividad AS DATE) as fecha_actividad,
    dias_despues_registro,
    activo
FROM user_activity;
'''
# Cargar datos con la consulta correcta
user_activity = pd.read_sql(query_user_activity_converted, con=engine)

# Convertir fecha en pandas
user_activity['fecha_actividad'] = pd.to_datetime(user_activity['fecha_actividad'])

# Verificar el resultado
user_activity.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32000 entries, 0 to 31999
Data columns (total 4 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   id_usuario             32000 non-null  object        
 1   fecha_actividad        32000 non-null  datetime64[ns]
 2   dias_despues_registro  32000 non-null  int64         
 3   activo                 32000 non-null  int64         
dtypes: datetime64[ns](1), int64(2), object(1)
memory usage: 1000.1+ KB


In [51]:
# Verificar si hay usuarios duplicados
query_duplicados = '''
SELECT id_usuario,
       count(fecha_registro) AS registros_duplicados
FROM users
GROUP BY id_usuario
HAVING count(fecha_registro) > 1
'''

duplicados = pd.read_sql(query_duplicados, con=engine)
print(f"Usuarios con registros duplicados: {len(duplicados)}")
duplicados.head()


Usuarios con registros duplicados: 0


,id_usuario,registros_duplicados


In [52]:
# Asignación id_usuario a cohorte
query_cohort_fecha = '''
SELECT id_usuario,
MIN(CAST(fecha_registro AS DATE)) AS cohort_fecha
FROM users
GROUP BY id_usuario
ORDER BY MIN(fecha_registro) ASC;
'''
resultado = pd.read_sql(query_cohort_fecha, con=engine)
resultado

,id_usuario,cohort_fecha
0,user_465,2025-01-01
1,user_4691,2025-01-01
2,user_5408,2025-01-01
3,user_5803,2025-01-01
4,user_4803,2025-01-01
...,...,...
7995,user_4287,2025-05-31
7996,user_6372,2025-05-31
7997,user_184,2025-05-31
7998,user_3762,2025-05-31


In [53]:
# Retención por cohortes
# ======================
query_cohort_retention_final = '''
WITH cohortes AS (
    SELECT
        id_usuario,
        DATE_TRUNC('month', CAST(fecha_registro AS DATE)) AS cohorte_mes
    FROM users

),

cohort_size AS (
    SELECT
        cohorte_mes,
        COUNT(DISTINCT id_usuario) AS clientes_iniciales
    FROM cohortes
    GROUP BY cohorte_mes

),

actividad_usuario AS (
    SELECT
        ua.id_usuario,
        c.cohorte_mes,
        MAX(CASE
            WHEN ua.dias_despues_registro >= 1
                 AND ua.activo = 1
            THEN 1 ELSE 0
        END) AS retained_w1,
        MAX(CASE
            WHEN ua.dias_despues_registro >= 8
                 AND ua.activo = 1
            THEN 1 ELSE 0
        END) AS retained_w2,
        MAX(CASE
            WHEN ua.dias_despues_registro >= 15
                 AND ua.activo = 1
            THEN 1 ELSE 0
        END) AS retained_w3
    FROM user_activity ua
    JOIN cohortes c
        ON ua.id_usuario = c.id_usuario
    GROUP BY
        ua.id_usuario,
        c.cohorte_mes
),

retention_summary AS (

    SELECT
        cohorte_mes,
        SUM(retained_w1) AS retained_w1,
        SUM(retained_w2) AS retained_w2,
        SUM(retained_w3) AS retained_w3
    FROM actividad_usuario
    GROUP BY cohorte_mes
)

SELECT
    TO_CHAR(cs.cohorte_mes, 'YYYY-MM') AS cohorte,
    cs.clientes_iniciales,
    rs.retained_w1,
    rs.retained_w2,
    rs.retained_w3,

    ROUND(rs.retained_w1 * 1.0 / cs.clientes_iniciales, 2) AS semana_1,
    ROUND(rs.retained_w2 * 1.0 / cs.clientes_iniciales, 2) AS semana_2,
    ROUND(rs.retained_w3 * 1.0 / cs.clientes_iniciales, 2) AS semana_3

FROM cohort_size cs

LEFT JOIN retention_summary rs
    ON cs.cohorte_mes = rs.cohorte_mes
ORDER BY cs.cohorte_mes;

'''
# Ejecutar la consulta
cohort_retention_final = pd.read_sql(query_cohort_retention_final, con=engine)
cohort_retention_final


,cohorte,clientes_iniciales,retained_w1,retained_w2,retained_w3,semana_1,semana_2,semana_3
0,2025-01,1627,1381,1253,1027,0.85,0.77,0.63
1,2025-02,1444,1255,1154,940,0.87,0.80,0.65
2,2025-03,1636,1428,1306,1060,0.87,0.80,0.65
3,2025-04,1606,1394,1261,1022,0.87,0.79,0.64
4,2025-05,1687,1446,1321,1088,0.86,0.78,0.64


In [54]:
# Análisis de cohortes
cohort_data = cohort_retention_final.copy()

# Exportar para Power BI
cohort_data.to_csv('cohort_retention_clean.csv', index=False)

**¿Los usuarios regresan?** Sí. La retención es sólida y consistente durante las primeras semanas.
Semana 1: 85%–87% de usuarios regresan.
Semana 2: 77%–80% continúan activos.
Semana 3: 63%–65% siguen utilizando la plataforma.

✅ La plataforma logra una fuerte activación inicial.
✅ Existe buena recurrencia temprana y estabilidad entre cohortes.
⚠️ La principal fuga ocurre después de la segunda semana.

📈 Insight estratégico
El negocio está reteniendo bien a los usuarios nuevos, pero la oportunidad está en fortalecer engagement y fidelización para sostener la actividad a largo plazo.

🏁 Conclusión
La adquisición funciona y los usuarios sí regresan.
El siguiente reto estratégico es reducir la caída después de las primeras semanas y aumentar la retención de largo plazo.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - El análisis del funnel conecta correctamente los resultados técnicos con implicaciones de negocio y experiencia de usuario. La interpretación de las etapas de conversión permite identificar con claridad el principal punto de abandono.
</div>


---

## 🔹 Paso 5: Validar si los cambios generan impacto (test estadístico)

🎯 **Objetivo:** Evaluar si la modificación en la UI del checkout impacta la **tasa de conversión de compra**.

---

1. **Analizar el dataset** `experiment_checkout_ui.csv` para identificar la métrica principal **conversion**.
   - La métrica **conversion** es 1 si el usuario completó la compra, 0 si no.    
2. **Plantear la hipótesis estadística**     
3. **Aplicar el test estadístico adecuado** 
4. **Interpretar el resultado**  

---
Hipótesis estadística
   - **H₀ (Hipótesis nula):** La nueva UI NO impacta las conversiones (no hay diferencia)
   - **H₁ (Hipótesis alternativa):** La nueva UI SÍ impacta las conversiones (hay diferencia)
   
**Test estadístico:** Chi-cuadrado: 0.6178 Valor bajo = poca diferencia entre control y tratamiento
Valor p: 0.4319
- Probabilidad de que la diferencia sea casualidad
- 43.19% de probabilidad de que sea solo suerte

Regla de decisión:
- Si p-value < α (0.05) → Rechazamos H₀ (hay diferencia significativa)
- Si p-value ≥ α (0.05) → No rechazamos H₀ (no hay diferencia significativa)
-**Nivel de significancia alpha:** α = 0.05 (95% confianza) - ESTÁNDAR "Con 95% de confianza, NO hay evidencia suficiente de que la nueva UI impacte las conversiones" 
Resultado: p-value (0.4319) > α (0.05)

Conclusión: ❌ No rechazamos H₀
- La nueva UI NO genera un impacto estadísticamente significativo en las conversiones
- La diferencia del 0.6% entre control/tratamiento probablemente es casualidad

💡 Respuesta final con los resultados
"Los resultados muestran que la nueva UI del checkout NO mejora significativamente las conversiones. Aunque el grupo de tratamiento tuvo una tasa ligeramente superior (16.29% vs 15.69%), esta diferencia no es estadísticamente significativa (p = 0.4319 > 0.05). Por tanto, no se recomienda implementar el cambio. El experimento demostró que los cambios propuestos en la UI del checkout no generaron una mejora significativa en las conversiones. 
Esto podría significar que:
* Quizás los cambios de la UI tendrían que estar más focalizados en el **add_payment_info** dado que fue detectado como el evento donde más se pierden clientes en el proceso de compra. 
* La UI actual ya funciona bien, y los cambios no fueron signficativos para resolver el problema de fondo del **add_payment_info**. 
* Los cambios probados no resolvieron problemas reales que están teniendo los usuarios.
* Se debe investigar otras áreas de mejora en el proceso de compra, probar otras UI con mayor precisión detectando antes cuáles son las dificultades de los usuarios.

In [55]:
# Librerías básicas
import pandas as pd
import numpy as np
from statsmodels.stats.proportion import proportions_ztest
from scipy.stats import chi2_contingency

In [56]:
# Cargar el dataset usando la URL directa
df = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/experiment_checkout_ui.csv')

# Exploración inicial
print("📊 Información del dataset:")
print(f"Filas: {df.shape[0]}, Columnas: {df.shape[1]}")
print("\n🔍 Primeras 5 filas:")
print(df.head())

📊 Información del dataset:
Filas: 10000, Columnas: 7

🔍 Primeras 5 filas:
   id_usuario     variante  convirtio dispositivo       pais  duracion_sesion  \
0  exp_user_0  tratamiento          0      mobile  Argentina           114.41   
1  exp_user_1  tratamiento          0     desktop     Mexico           170.03   
2  exp_user_2      control          1      mobile   Colombia           140.21   
3  exp_user_3  tratamiento          0      mobile   Colombia           151.45   
4  exp_user_4  tratamiento          0     desktop     Mexico           299.96   

    timestamp  
0  2025-03-28  
1  2025-01-15  
2  2025-03-18  
3  2025-06-03  
4  2025-01-12  


In [57]:
# Verificar la distribución de variantes
print("Distribución de variantes:")
print(df['variante'].value_counts())
print("\nPorcentaje por grupo:")
print(df['variante'].value_counts(normalize=True) * 100)

Distribución de variantes:
tratamiento    5035
control        4965
Name: variante, dtype: int64

Porcentaje por grupo:
tratamiento    50.35
control        49.65
Name: variante, dtype: float64


In [58]:
# Analizar distribución de la variable de conversión
print("Distribución de conversiones:")
print(df['convirtio'].value_counts())
print("\nPorcentaje de conversiones:")
print(df['convirtio'].value_counts(normalize=True) * 100)

Distribución de conversiones:
0    8401
1    1599
Name: convirtio, dtype: int64

Porcentaje de conversiones:
0    84.01
1    15.99
Name: convirtio, dtype: float64


In [59]:
# Verificar conversiones por variante renombrar columnas
print("\n Conversiones por variante:")
conversion_summary = df.groupby('variante')['convirtio'].agg(['count', 'sum', 'mean'])
conversion_summary.columns = ['Total_usuarios', 'Conversiones', 'Tasa_conversion']
conversion_summary['Tasa_conversion_pct'] = conversion_summary['Tasa_conversion'] * 100
print(conversion_summary)


 Conversiones por variante:
             Total_usuarios  Conversiones  Tasa_conversion  \
variante                                                     
control                4965           779         0.156898   
tratamiento            5035           820         0.162860   

             Tasa_conversion_pct  
variante                          
control                15.689829  
tratamiento            16.285998  


In [60]:
# Extraer conversiones y totales por variante
control_conversions = 779
treatment_conversions = 820
control_total = 4965
treatment_total = 5035

from scipy.stats import chi2_contingency
tabla_contingencia = np.array([
    [treatment_conversions, treatment_total - treatment_conversions],
    [control_conversions, control_total - control_conversions]
])

chi2_stat, p_value, dof, expected = chi2_contingency(tabla_contingencia)

# VER RESULTADOS:
print("=== RESULTADOS DEL TEST ESTADÍSTICO ===")
print(f"Estadístico Chi-cuadrado: {chi2_stat:.4f}")
print(f"Valor p: {p_value:.4f}")
print(f"Grados de libertad: {dof}")
print(f"Nivel de significancia α: 0.05")

# Interpretación
if p_value < 0.05:
    print("✅ RESULTADO: Diferencia estadísticamente significativa")
    print("   Rechazamos H₀ - La nueva UI SÍ impacta las conversiones")
else:
    print("❌ RESULTADO: No hay diferencia estadísticamente significativa")
    print("   No rechazamos H₀ - La nueva UI NO impacta las conversiones")


=== RESULTADOS DEL TEST ESTADÍSTICO ===
Estadístico Chi-cuadrado: 0.6178
Valor p: 0.4319
Grados de libertad: 1
Nivel de significancia α: 0.05
❌ RESULTADO: No hay diferencia estadísticamente significativa
   No rechazamos H₀ - La nueva UI NO impacta las conversiones


In [61]:
# Crear dataset con resultados del experimento A/B
experiment_results = pd.DataFrame({
    'variante': ['control', 'tratamiento'],
    'total_usuarios': [4965, 5035],
    'conversiones': [779, 820],
    'tasa_conversion': [15.69, 16.29],
    'chi2_stat': [0.6178, 0.6178],
    'p_value': [0.4319, 0.4319],
    'significativo': ['No', 'No']
})

# Exportar para Power BI
experiment_results.to_csv('experiment_ab_results.csv', index=False)

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - El análisis de cohortes está bien estructurado y permite interpretar claramente el comportamiento de retención de los usuarios. La conexión entre resultados técnicos e insights estratégicos fortalece el valor del análisis para el negocio.
</div>


---

## 🔹 Paso 6: Comunicar los resultados (Dashboard en BI)

🎯 **Objetivo**:  
Crear un dashboard que muestre de manera clara y visual los resultados del análisis de ventas, costos, marketing y conversión. 

Se usarán los CSVs limpios del Paso 1:

- `orders_clean.csv`  
- `catalog_clean.csv`  
- `marketing_clean.csv`

---

1️⃣ Preparación de los datos
1. Cargar los CSVs en Power BI o Tableau.
2. Revisar relaciones:
   - `orders.nombre_producto` → `catalog.nombre_producto`
   - `orders.fecha_pedido` → tabla de fechas (crear calendario para análisis temporal)
   - `orders.fecha_pedido` → `dim_fecha.date`
3. Crear columnas calculadas necesarias
4. Crear tabla de fechas para poder calcular comparaciones YTD, YoY o períodos anteriores (`Previous Year`, `Previous Month`).

---

2️⃣ Dashboard 1: O>
**KPIs principales a mostrar:**
- Revenue total
- Profit total
- Gasto total en marketing
- Ticket promedio
- Cantidad promedio de productos por orden

**Visualizaciones sugeridas:**
- Tarjetas KPI para revenue, profit y gasto marketing
- Gráfico de líneas: evolución mensual de revenue o profit
- Gráfico de líneas YTD
- Gráfico de barras: revenue y profit por producto o categoría

---

 3️⃣ Dashboard 2: Detalle / Drill-through  
**Objetivo:** Permitir explorar los datos desde el KPI general hasta cada orden o producto.

**Visualizaciones sugeridas:**
- Tabla detallada de órdenes con:
  - producto, cantidad, revenue, cost, profit
  - color condicional (profit negativo en rojo, positivo en verde)
- Gráfico de barras por producto con medida `cantidad vendida`
- Drill-through: seleccionar un producto y ver todos los pedidos relacionados
- Filtros por fecha, categoría de producto, etc

---

## 🚀 Entrega Final

Comparte el acceso a tu Dashboard para revisión.   
Puedes entregar el Dashboard utilizando **Power BI o Tableau**.

Incluye **uno de los siguientes**:

- 🔗 Link público del dashboard publicado en **Power BI Service o Tableau Public / Tableau Cloud**
- 🔗 Link de **Google Drive o OneDrive** con el archivo del proyecto (`.pbix`) y los 3 csvs limpios.


### 📎 Enlace del Dashboard

In [62]:
# (Pega aquí tu link)
# link de one drive / google drive

(RENVÍO ACTUALIZACIÓN DEL LINK)
https://drive.google.com/drive/folders/1QKJeqRy0gOpTg-3W3-USIcNwEE-C1TTx?usp=sharing



<S>
<div class="alert alert-block alert-warning">

<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Atención</b> ⚠️ - Gracias por compartir el enlace al dashboard y los recursos del proyecto. Sin embargo, actualmente el link de Google Drive parece dirigir únicamente a un archivo CSV y no al dashboard o archivo interactivo esperado. 

Te recomendaría verificar nuevamente el enlace compartido y confirmar que:
<ul>
<li>El archivo correcto del dashboard fue subido</li>

<li>Los permisos de acceso estén habilitados para visualización</li>

<li>El enlace corresponda al dashboard/exportación final y no solo a los datos fuente</li>
</ul>

Compartir correctamente el dashboard final es muy importante para que el lector pueda explorar la visualización interactiva y validar completamente el análisis realizado. ¡El trabajo visual se ve bastante sólido, así que vale mucho la pena asegurarse de que pueda accederse correctamente! 🚀📊

</div>

Hola equipo revisor TripleTen,  
Adjunto la entrega del proyecto final con las 2 hojas del dashboard, esperando cubrir los propósitos solicitados. Y los 3 datasets limpios.  
De forma adicional, comparto un archivo Anexo con 3 vistas exploratorias que desarrollé durante el análisis. Las dejo a su consideración por si son de utilidad para contexto y desarrollo del proyecto.  
Quedo atenta a sus comentarios.  Gracias,  
Ana Delia :)

<s>

# Comentario General del Revisor

<div class="alert alert-block alert-warning">

<b>Comentario del revisor</b> <a class="tocSkip"></a>

¡Excelente trabajo en este proyecto final! 
El proyecto demuestra una integración muy sólida de análisis de datos, SQL, estadística y Business Intelligence dentro de un flujo analítico bien estructurado y orientado a negocio. Se nota mucho esfuerzo tanto en la parte técnica como en la comunicación visual de resultados.

Destaco especialmente:
<ul>
<li>La calidad y organización del análisis exploratorio</li>

<li>La construcción de KPIs relevantes para negocio</li>

<li>La correcta aplicación de pruebas estadísticas</li>

<li>La estructura profesional de los dashboards y el storytelling analítico</li>

<li>La integración entre Python, SQL y visualización</li>
</ul>


En general, el proyecto está muy bien desarrollado y bastante cercano a entregables reales utilizados en entornos profesionales de analytics y BI. 

Actualmente, el único detalle pendiente es revisar el enlace compartido del dashboard, ya que el link de Google Drive parece dirigir únicamente a un archivo CSV y no al dashboard interactivo final. Una vez corregido ese punto y compartida correctamente la visualización final, el proyecto quedará completamente cerrado. 

¡Gran trabajo en este proyecto final y felicitaciones por todo el esfuerzo invertido!

</div>

Gracias Hesus, por tus comentarios. Sin duda nunca olvidaré que fuiste el primer con H. Quedo atenta a tu revisión, esperando que el link esta vez funcione correctamente. Saludos :) 

<div class="alert alert-block alert-info">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

Gracias por considerar las observaciones realizadas y aplicar las mejoras sugeridas en el proyecto. Se aprecia el trabajo dedicado para fortalecer tanto el análisis con la presentación del dashboard.



</div>




# Comentario General del Revisor

<div class="alert alert-block alert-success">

<b>Comentario del revisor</b> <a class="tocSkip"></a>

¡Excelente trabajo en este proyecto final!
El desarrollo completo refleja una integración sólida entre análisis de datos, SQL, estadística y visualización orientada a negocio. El flujo analítico mantiene coherencia desde la limpieza de datos hasta la construcción del dashboard ejecutivo, permitiendo responder claramente las preguntas planteadas.

<br>

<b>Evaluación del código y análisis</b>

Destaco especialmente:

<ul>
<li>La limpieza y validación detallada de los datasets, incluyendo tratamiento de nulos, revisión de duplicados, validación de variables numéricas y estandarización de categorías.</li>

<li>La correcta construcción de KPIs financieros como revenue, costo total, profit y gasto de marketing, permitiendo evaluar claramente la rentabilidad del negocio.</li>

<li>La integración eficiente de múltiples datasets mediante <code>merge</code> para enriquecer el análisis comercial y financiero.</li>

<li>La implementación estructurada de consultas SQL para el análisis de funnel y cohortes, utilizando CTEs y lógica de agregación de manera organizada.</li>

<li>La interpretación adecuada del funnel de conversión, identificando visualmente las etapas donde se presenta la mayor pérdida de usuarios.</li>

<li>La construcción correcta del análisis de retención semanal por cohortes, mostrando una lectura clara del comportamiento de usuarios recurrentes.</li>

<li>La aplicación apropiada del Z-Test para validar diferencias entre variantes del experimento de checkout, incluyendo hipótesis, nivel de significancia e interpretación estadística.</li>

<li>La organización general del notebook facilita la lectura y mantiene una estructura profesional durante todo el proyecto.</li>
</ul>

<br>

<b>Evaluación del dashboard</b>

El dashboard presenta una estructura visual profesional y orientada a la toma de decisiones. La navegación entre páginas permite pasar de una visión ejecutiva general hacia un análisis más operativo y detallado de productos, revenue y rentabilidad.

Se destacan especialmente:

<ul>
<li>La correcta presentación de KPIs ejecutivos como Revenue Total, Profit Total, Costo Total, Ticket Promedio y Gasto de Marketing, permitiendo una lectura inmediata del desempeño del negocio.</li>

<li>El gráfico de tendencia temporal comunica claramente cómo evolucionan revenue y profit entre enero y junio, facilitando detectar variaciones importantes en la rentabilidad.</li>

<li>La incorporación de conclusiones narrativas dentro del dashboard fortalece el storytelling analítico y conecta directamente los datos con hallazgos de negocio.</li>

<li>La comparación de revenue por país permite identificar rápidamente que Argentina y México concentran la mayor generación de ingresos.</li>

<li>La visualización por canal muestra adecuadamente el impacto comercial de organic y social frente a paid_search.</li>

<li>La matriz “Profit vs Revenue” aporta una lectura estratégica sobre productos con alto volumen pero baja rentabilidad, ayudando a detectar oportunidades de optimización.</li>

<li>La tabla transaccional con órdenes de profit negativo agrega profundidad analítica y permite identificar casos críticos específicos para revisión.</li>

<li>La evaluación de rentabilidad por dispositivo aporta valor de negocio al detectar diferencias claras entre mobile y desktop.</li>

<li>El uso de filtros dinámicos mejora significativamente la exploración interactiva de la información.</li>

<li>La coherencia visual, distribución de elementos y jerarquía de información generan un dashboard limpio, entendible y profesional.</li>
</ul>

El proyecto logra transformar datos complejos en conclusiones claras y accionables, mostrando una visión analítica bien desarrollada y alineada con necesidades reales de negocio.

</div>
